### PR TITLE
Document System.IO.Compression BrotliEncoder and BrotliStream

### PR DESCRIPTION
**Pull request by carlossanlop**
Here are all the related issues/proposals/epics/blogposts/PRs I found for these APIs:

https://devblogs.microsoft.com/dotnet/introducing-support-for-brotli-compression/
Add Brotli Compression to CoreFX - https://github.com/dotnet/corefx/pull/26229
System.IO.Compression.Brotli Epic - https://github.com/dotnet/corefx/issues/24826
Pure managed implementation of System.IO.Compression.Brotli - https://github.com/dotnet/corefxlab/issues/1673
Add Brotli Compression to CoreFX - https://github.com/dotnet/corefx/issues/25785

@ahsonkhan @ViktorHofer co-owners.
@ahsonkhan I noticed you were heavily involved in the PRs that introduced the code. Can you please help review these changes?

### CODE REVIEW

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 1]
```diff
@@ -142,10 +151,10 @@
         <Parameter Name="bytesWritten" Type="System.Int32" RefType="out" />
       </Parameters>
       <Docs>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <summary>Compresses an empty read-only span of bytes into <paramref name="destination" />, ensuring output is produced for all the processed input. An actual flush is performed when the source is depleted and there is enough space in <paramref name="destination" />.</summary>
```
Comment (carlossanlop) [inline xml/System.IO.Compression/BrotliEncoder.xml]: The [source code](https://github.com/dotnet/corefx/blob/c668e96cb0886d13a051d8021a3c3147e7daa15c/src/System.IO.Compression.Brotli/src/System/IO/Compression/enc/BrotliEncoder.cs#L113) for `Flush` internally calls `Compress` with an empty source parameter. I wrote a very literal description of what this method does, but I don't really understand the true purpose. Can you please help me improve this description, @ahsonkhan?

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 2]
```diff
@@ -22,7 +22,7 @@
     </Interface>
   </Interfaces>
   <Docs>
-    <summary>To be added.</summary>
+    <summary>Provides methods and static methods to encode and decode data in a streamless, non-allocating and performant manner using the Brotli data format specification.</summary>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
    <summary>Provides methods and static methods to encode and decode data in a streamless, non-allocating, and performant manner using the Brotli data format specification.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 3]
```diff
@@ -47,10 +47,19 @@
         <Parameter Name="window" Type="System.Int32" />
       </Parameters>
       <Docs>
-        <param name="quality">To be added.</param>
-        <param name="window">To be added.</param>
-        <summary>To be added.</summary>
+        <param name="quality">A number representing quality of the Brotli compression. 0 is the minimum (no compression), 11 is the maximum.</param>
+        <param name="window">A number representing the encoder window bits. The minimum value is 10, and the maximum value is 24.</param>
+        <summary>Initializes an instance of <see cref="T:System.IO.Compression.BrotliEncoder" /> class using the specified quality and window.</summary>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <summary>Initializes a new instance of the <see cref="T:System.IO.Compression.BrotliEncoder" /> structure using the specified quality and window.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 4]
```diff
@@ -47,10 +47,19 @@
...
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="quality" /> is not between the minimum value of 0 and the maximum value of 11.
+
+-or-
+
+<paramref name="window" /> is not between the minimum value of 10 and the maximum value of 24.
+        </exception>
+        <exception cref="T:System.InvalidOperationException">Quality and window cannot be changed at the current encoder state.</exception>
+        <exception cref="T:System.ObjectDisposedException">Cannot access a closed encoder.</exception>
+        <exception cref="T:System.IOException">Failed to create the <see cref="T:System.IO.Compression.BrotliEncoder" /> instance.</exception>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <exception cref="T:System.IO.IOException">Failed to create the <see cref="T:System.IO.Compression.BrotliEncoder" /> instance.</exception>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 5]
```diff
@@ -47,10 +47,19 @@
...
+        <param name="window">A number representing the encoder window bits. The minimum value is 10, and the maximum value is 24.</param>
+        <summary>Initializes an instance of <see cref="T:System.IO.Compression.BrotliEncoder" /> class using the specified quality and window.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="quality" /> is not between the minimum value of 0 and the maximum value of 11.
+
+-or-
+
+<paramref name="window" /> is not between the minimum value of 10 and the maximum value of 24.
+        </exception>
+        <exception cref="T:System.InvalidOperationException">Quality and window cannot be changed at the current encoder state.</exception>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]: This and the next exception are surprising in a constructor. Instantiation should initialize the encoder state. And there shouldn't be a call to Dispose before the instantiation.

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 6]
```diff
@@ -80,13 +89,13 @@
...
-        <param name="bytesWritten">To be added.</param>
-        <param name="isFinalBlock">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesConsumed">When this method returns, the total number of bytes that were read from <paramref name="source" />.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <param name="isFinalBlock"><see langword="true" /> to finalize the internal stream, preventing adding more input data when this method returns; <see langword="false" /> to allow the encoder to postpone the production of output until it has processed enough input.</param>
+        <summary>Compresses <paramref name="source" /> into <paramref name="destination" />.</summary>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <summary>Compresses a read-only byte span into a destination span.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 7]
```diff
@@ -80,13 +89,13 @@
...
       <Docs>
-        <param name="source">To be added.</param>
-        <param name="destination">To be added.</param>
-        <param name="bytesConsumed">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <param name="isFinalBlock">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <param name="destination">When this method returns, a byte span where the compressed is stored.</param>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 8]
```diff
@@ -80,13 +89,13 @@
...
+        <param name="bytesConsumed">When this method returns, the total number of bytes that were read from <paramref name="source" />.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <param name="isFinalBlock"><see langword="true" /> to finalize the internal stream, preventing adding more input data when this method returns; <see langword="false" /> to allow the encoder to postpone the production of output until it has processed enough input.</param>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <param name="isFinalBlock"><see langword="true" /> to finalize the internal stream, which prevents adding more input data when this method returns; <see langword="false" /> to allow the encoder to postpone the production of output until it has processed enough input.</param>
```

File: xml/System.IO.Compression/BrotliEncoder.xml (hunk 1 above)
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <summary>Compresses an empty read-only span of bytes into its destination, which ensures that output is produced for all the processed input. An actual flush is performed when the source is depleted and there is enough space in the destination for the remaining data.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 9]
```diff
@@ -142,10 +151,10 @@
...
       </Parameters>
       <Docs>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <summary>Compresses an empty read-only span of bytes into <paramref name="destination" />, ensuring output is produced for all the processed input. An actual flush is performed when the source is depleted and there is enough space in <paramref name="destination" />.</summary>
+        <returns>One of the enumeration values that describes the status with which the span-based operation finished.</returns>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <returns>One of the enumeration values that describes the status with which the operation finished.</returns>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 10]
```diff
@@ -203,11 +221,11 @@
...
       <Docs>
-        <param name="source">To be added.</param>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <summary>Tries to compress <paramref name="source" /> into <paramref name="destination" />.</summary>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <summary>Tries to compress a source byte span into a destination span.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 11]
```diff
@@ -203,11 +221,11 @@
         <Parameter Name="bytesWritten" Type="System.Int32" RefType="out" />
       </Parameters>
       <Docs>
-        <param name="source">To be added.</param>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <param name="destination">When this method returns, a span of bytes where the compressed data is stored.</param>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 12]
```diff
@@ -238,13 +256,13 @@
...
-        <param name="quality">To be added.</param>
-        <param name="window">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <param name="quality">A number representing quality of the Brotli compression. 0 is the minimum (no compression), 11 is the maximum.</param>
+        <param name="window">A number representing the encoder window bits. The minimum value is 10, and the maximum value is 24.</param>
+        <summary>Tries to compress <paramref name="source" /> into <paramref name="destination" />, using the provided compression quality leven and encoder window bits.</summary>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <summary>Tries to compress a source byte span into a destination byte span, using the provided compression quality leven and encoder window bits.</summary>
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 13]
```diff
@@ -238,13 +256,13 @@
...
       <Docs>
-        <param name="source">To be added.</param>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <param name="quality">To be added.</param>
-        <param name="window">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="source">A read-only span of bytes containing the source data to compress.</param>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
```
Comment (rpetrusha) [inline xml/System.IO.Compression/BrotliEncoder.xml]:
```suggestion
        <param name="destination">When this method returns, a span of bytes where the compressed data is stored.</param>
```

File: xml/System.IO.Compression/BrotliEncoder.xml (hunk 5 above)
Comment (carlossanlop) [reply xml/System.IO.Compression/BrotliEncoder.xml]: You're right, this is interesting. I followed the constructor code and reached these two points, but I am not sure if when creating a new instance we would ever reach those exceptions. Let me review this code again, see if there's a chance of hitting the two exceptions. But for now, I'll remove them:
```suggestion
```

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 14]
```diff
@@ -47,10 +47,19 @@
...
+        <summary>Initializes an instance of <see cref="T:System.IO.Compression.BrotliEncoder" /> class using the specified quality and window.</summary>
         <remarks>To be added.</remarks>
+        <exception cref="T:System.ArgumentOutOfRangeException"><paramref name="quality" /> is not between the minimum value of 0 and the maximum value of 11.
+
+-or-
+
+<paramref name="window" /> is not between the minimum value of 10 and the maximum value of 24.
+        </exception>
+        <exception cref="T:System.InvalidOperationException">Quality and window cannot be changed at the current encoder state.</exception>
+        <exception cref="T:System.ObjectDisposedException">Cannot access a closed encoder.</exception>
```
Comment (carlossanlop) [inline xml/System.IO.Compression/BrotliEncoder.xml]: Removed based on @rpetrusha 's above comment. Will investigate if it's truly possible to throw this exception in the constructor.
```suggestion
```

**Comment by rpetrusha** [comment]
I'll merge now, @carlossanlop.

File: xml/System.IO.Compression/BrotliEncoder.xml [hunk 15]
```diff
@@ -142,10 +149,10 @@
         <Parameter Name="bytesWritten" Type="System.Int32" RefType="out" />
       </Parameters>
       <Docs>
-        <param name="destination">To be added.</param>
-        <param name="bytesWritten">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="destination">When this method returns, a span of bytes where the compressed data will be stored.</param>
+        <param name="bytesWritten">When this method returns, the total number of bytes that were written to <paramref name="destination" />.</param>
+        <summary>Compresses an empty read-only span of bytes into its destination, which ensures that output is produced for all the processed input. An actual flush is performed when the source is depleted and there is enough space in the destination for the remaining data.</summary>
```
Comment (ahsonkhan) [inline xml/System.IO.Compression/BrotliEncoder.xml]: `Compresses an empty read-only span of bytes into its destination`

This is an implementation detail that is not relevant to the caller (and hence shouldn't be documented).

**Pull request merged**